{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Script to prepare LD reference data for each ancestry group  \n",
    "\n",
    "After downloading and processing 1000 genomes project LD panels with `1.download_and_prep_1kg_ld_panels.sh`, this `Julia` script prepares LD reference panels from the 1000 genomes project for each ancestry group as follows  \n",
    "- keep SNPs with minor allele frequency > 0.05\n",
    "- keep polymorphic SNPs\n",
    "\n",
    "After filtering, the sciript keeps the filtered .bed file for each ancestry group and removes any intermediate files. The script does not remove the original vcf or plink files.  \n",
    "\n",
    "Julia v.1.8.1 was used for this script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir = \"/u/project/gandalm/shared/refGenomes/1000genomes_GeneticsMakie_chrpos\"\n",
    "cd(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m registry at `~/.julia/registries/General.toml`\n",
      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.8/Project.toml`\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.8/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "using Pkg\n",
    "Pkg.add([\"Downloads\", \"SnpArrays\", \"CSV\", \"DataFrames\", \"Glob\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "using Downloads, SnpArrays, CSV, DataFrames, Glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "true"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isdir(\"data_ancestry\") || mkdir(\"data_ancestry\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing filtered .bed files for each ancestry group"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download sample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "true"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download sample metadata\n",
    "beagle = \"https://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a\"\n",
    "url = joinpath(beagle, \"sample_info/integrated_call_samples_v3.20130502.ALL.panel\")\n",
    "meta = last(split(url, \"/\")) \n",
    "isfile(\"data_ancestry/$(meta)\") || Downloads.download(url, \"data_ancestry/$(meta)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"data-frame\"><p>3 rows × 4 columns</p><table class=\"data-frame\"><thead><tr><th></th><th>sample</th><th>pop</th><th>super_pop</th><th>gender</th></tr><tr><th></th><th title=\"String7\">String7</th><th title=\"String3\">String3</th><th title=\"String3\">String3</th><th title=\"String7\">String7</th></tr></thead><tbody><tr><th>1</th><td>HG00096</td><td>GBR</td><td>EUR</td><td>male</td></tr><tr><th>2</th><td>HG00097</td><td>GBR</td><td>EUR</td><td>female</td></tr><tr><th>3</th><td>HG00099</td><td>GBR</td><td>EUR</td><td>female</td></tr></tbody></table></div>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& sample & pop & super\\_pop & gender\\\\\n",
       "\t\\hline\n",
       "\t& String7 & String3 & String3 & String7\\\\\n",
       "\t\\hline\n",
       "\t1 & HG00096 & GBR & EUR & male \\\\\n",
       "\t2 & HG00097 & GBR & EUR & female \\\\\n",
       "\t3 & HG00099 & GBR & EUR & female \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "\u001b[1m3×4 DataFrame\u001b[0m\n",
       "\u001b[1m Row \u001b[0m│\u001b[1m sample  \u001b[0m\u001b[1m pop     \u001b[0m\u001b[1m super_pop \u001b[0m\u001b[1m gender  \u001b[0m\n",
       "\u001b[1m     \u001b[0m│\u001b[90m String7 \u001b[0m\u001b[90m String3 \u001b[0m\u001b[90m String3   \u001b[0m\u001b[90m String7 \u001b[0m\n",
       "─────┼──────────────────────────────────────\n",
       "   1 │ HG00096  GBR      EUR        male\n",
       "   2 │ HG00097  GBR      EUR        female\n",
       "   3 │ HG00099  GBR      EUR        female"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = CSV.read(\"data_ancestry/integrated_call_samples_v3.20130502.ALL.panel\", DataFrame, header=false, skipto = 2)\n",
    "rename!(meta, [:sample, :pop, :super_pop, :gender])\n",
    "first(meta, 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Filter SNPs with maf > 0.05 and mac > 1  \n",
    "Using the processed 1000 genomes project LD panels (see `1.download_and_prep_1kg_ld_panels.sh` for the process), this section will take care of the following things:\n",
    "1. subset to an ancestry group\n",
    "2. filter out SNPs with minor allele frequency < 0.05\n",
    "3. filter out SNPs with minor allele count < 1\n",
    "\n",
    "note: The logs directory stores the log for this section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2504, 32504336)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basefile = \"kgp\"\n",
    "kgp_original = SnpData(\"$(dir)/$(basefile)\")\n",
    "size_kgp_original = size(kgp_original) # to be used to compare before/after the filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "### processing AFR samples ###\n",
      " \n",
      "### subsetting to AFR samples ###\n",
      "sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      " \n",
      "filtering SNPs with maf > 0.05\n",
      "dim for AFR file (before maf > 0.05)\n",
      "(661, 32504336)\n",
      "dim for AFR file (after maf > 0.05)\n",
      "(661, 10078577)\n",
      " \n",
      "filtering SNPs with mac > 1\n",
      "number of SNPs with mac > 1 (will be kept): 10078529\n",
      "dim for AFR file (after maf > 0.05, mac > 1)\n",
      "(661, 10078529)\n",
      "cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      "cf: dim original - (2504, 32504336)\n",
      " \n",
      "SnpData(people: 661, snps: 10078529,\n",
      "snp_info: \n",
      " Row │ chromosome  snpid         genetic_distance  position  allele1  allele2\n",
      "     │ String      String        Float64           Int64     String   String\n",
      "─────┼────────────────────────────────────────────────────────────────────────\n",
      "   1 │ 1           1:10177:A:AC               0.0     10177  AC       A\n",
      "   2 │ 1           1:10352:T:TA               0.0     10352  TA       T\n",
      "   3 │ 1           1:11008:C:G                0.0     11008  G        C\n",
      "   4 │ 1           1:11012:C:G                0.0     11012  G        C\n",
      "   5 │ 1           1:14599:T:A                0.0     14599  A        T\n",
      "   6 │ 1           1:14604:A:G                0.0     14604  G        A\n",
      "…,\n",
      "person_info: \n",
      " Row │ fid        iid        father     mother     sex        phenotype\n",
      "     │ Abstract…  Abstract…  Abstract…  Abstract…  Abstract…  Abstract…\n",
      "─────┼──────────────────────────────────────────────────────────────────\n",
      "   1 │ 0          HG01879    0          0          1          -9\n",
      "   2 │ 0          HG01880    0          0          2          -9\n",
      "   3 │ 0          HG01882    0          0          1          -9\n",
      "   4 │ 0          HG01883    0          0          2          -9\n",
      "   5 │ 0          HG01885    0          0          1          -9\n",
      "   6 │ 0          HG01886    0          0          2          -9\n",
      "…,\n",
      "srcbed: data_ancestry/kgp.AFR.maf005.mac1.bed\n",
      "srcbim: data_ancestry/kgp.AFR.maf005.mac1.bim\n",
      "srcfam: data_ancestry/kgp.AFR.maf005.mac1.fam\n",
      ")\n",
      " \n",
      " \n",
      "files to be removed: \n",
      "8\n",
      "Any[\"data_ancestry/kgp.AFR.bed\", \"data_ancestry/kgp.AFR.bim\", \"data_ancestry/kgp.AFR.fam\", \"data_ancestry/kgp.AFR.maf005.bed\", \"data_ancestry/kgp.AFR.maf005.bim\", \"data_ancestry/kgp.AFR.maf005.fam\", \"data_ancestry/kgp.AFR.maf005.mac1.fam\", \"data_ancestry/kgp.AFR.maf005.mac1.bim\"]\n",
      " \n",
      " \n",
      "### processing AMR samples ###\n",
      " \n",
      "### subsetting to AMR samples ###\n",
      "sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      " \n",
      "filtering SNPs with maf > 0.05\n",
      "dim for AMR file (before maf > 0.05)\n",
      "(347, 32504336)\n",
      "dim for AMR file (after maf > 0.05)\n",
      "(347, 7453323)\n",
      " \n",
      "filtering SNPs with mac > 1\n",
      "number of SNPs with mac > 1 (will be kept): 7453217\n",
      "dim for AMR file (after maf > 0.05, mac > 1)\n",
      "(347, 7453217)\n",
      "cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      "cf: dim original - (2504, 32504336)\n",
      " \n",
      "SnpData(people: 347, snps: 7453217,\n",
      "snp_info: \n",
      " Row │ chromosome  snpid         genetic_distance  position  allele1  allele2\n",
      "     │ String      String        Float64           Int64     String   String\n",
      "─────┼────────────────────────────────────────────────────────────────────────\n",
      "   1 │ 1           1:10177:A:AC               0.0     10177  AC       A\n",
      "   2 │ 1           1:10352:T:TA               0.0     10352  TA       T\n",
      "   3 │ 1           1:11008:C:G                0.0     11008  G        C\n",
      "   4 │ 1           1:11012:C:G                0.0     11012  G        C\n",
      "   5 │ 1           1:13116:T:G                0.0     13116  G        T\n",
      "   6 │ 1           1:13118:A:G                0.0     13118  G        A\n",
      "…,\n",
      "person_info: \n",
      " Row │ fid        iid        father     mother     sex        phenotype\n",
      "     │ Abstract…  Abstract…  Abstract…  Abstract…  Abstract…  Abstract…\n",
      "─────┼──────────────────────────────────────────────────────────────────\n",
      "   1 │ 0          HG00551    0          0          2          -9\n",
      "   2 │ 0          HG00553    0          0          1          -9\n",
      "   3 │ 0          HG00554    0          0          2          -9\n",
      "   4 │ 0          HG00637    0          0          1          -9\n",
      "   5 │ 0          HG00638    0          0          2          -9\n",
      "   6 │ 0          HG00640    0          0          1          -9\n",
      "…,\n",
      "srcbed: data_ancestry/kgp.AMR.maf005.mac1.bed\n",
      "srcbim: data_ancestry/kgp.AMR.maf005.mac1.bim\n",
      "srcfam: data_ancestry/kgp.AMR.maf005.mac1.fam\n",
      ")\n",
      " \n",
      " \n",
      "files to be removed: \n",
      "8\n",
      "Any[\"data_ancestry/kgp.AMR.bed\", \"data_ancestry/kgp.AMR.bim\", \"data_ancestry/kgp.AMR.fam\", \"data_ancestry/kgp.AMR.maf005.bed\", \"data_ancestry/kgp.AMR.maf005.bim\", \"data_ancestry/kgp.AMR.maf005.fam\", \"data_ancestry/kgp.AMR.maf005.mac1.fam\", \"data_ancestry/kgp.AMR.maf005.mac1.bim\"]\n",
      " \n",
      " \n",
      "### processing EAS samples ###\n",
      " \n",
      "### subsetting to EAS samples ###\n",
      "sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      " \n",
      "filtering SNPs with maf > 0.05\n",
      "dim for EAS file (before maf > 0.05)\n",
      "(504, 32504336)\n",
      "dim for EAS file (after maf > 0.05)\n",
      "(504, 6680335)\n",
      " \n",
      "filtering SNPs with mac > 1\n",
      "number of SNPs with mac > 1 (will be kept): 6680238\n",
      "dim for EAS file (after maf > 0.05, mac > 1)\n",
      "(504, 6680238)\n",
      "cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      "cf: dim original - (2504, 32504336)\n",
      " \n",
      "SnpData(people: 504, snps: 6680238,\n",
      "snp_info: \n",
      " Row │ chromosome  snpid         genetic_distance  position  allele1  allele2\n",
      "     │ String      String        Float64           Int64     String   String\n",
      "─────┼────────────────────────────────────────────────────────────────────────\n",
      "   1 │ 1           1:10177:A:AC               0.0     10177  AC       A\n",
      "   2 │ 1           1:10352:T:TA               0.0     10352  TA       T\n",
      "   3 │ 1           1:13273:G:C                0.0     13273  C        G\n",
      "   4 │ 1           1:14599:T:A                0.0     14599  A        T\n",
      "   5 │ 1           1:14604:A:G                0.0     14604  G        A\n",
      "   6 │ 1           1:14930:A:G                0.0     14930  G        A\n",
      "…,\n",
      "person_info: \n",
      " Row │ fid        iid        father     mother     sex        phenotype\n",
      "     │ Abstract…  Abstract…  Abstract…  Abstract…  Abstract…  Abstract…\n",
      "─────┼──────────────────────────────────────────────────────────────────\n",
      "   1 │ 0          HG00403    0          0          1          -9\n",
      "   2 │ 0          HG00404    0          0          2          -9\n",
      "   3 │ 0          HG00406    0          0          1          -9\n",
      "   4 │ 0          HG00407    0          0          2          -9\n",
      "   5 │ 0          HG00409    0          0          1          -9\n",
      "   6 │ 0          HG00410    0          0          2          -9\n",
      "…,\n",
      "srcbed: data_ancestry/kgp.EAS.maf005.mac1.bed\n",
      "srcbim: data_ancestry/kgp.EAS.maf005.mac1.bim\n",
      "srcfam: data_ancestry/kgp.EAS.maf005.mac1.fam\n",
      ")\n",
      " \n",
      " \n",
      "files to be removed: \n",
      "8\n",
      "Any[\"data_ancestry/kgp.EAS.bed\", \"data_ancestry/kgp.EAS.bim\", \"data_ancestry/kgp.EAS.fam\", \"data_ancestry/kgp.EAS.maf005.bed\", \"data_ancestry/kgp.EAS.maf005.bim\", \"data_ancestry/kgp.EAS.maf005.fam\", \"data_ancestry/kgp.EAS.maf005.mac1.fam\", \"data_ancestry/kgp.EAS.maf005.mac1.bim\"]\n",
      " \n",
      " \n",
      "### processing EUR samples ###\n",
      " \n",
      "### subsetting to EUR samples ###\n",
      "sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      " \n",
      "filtering SNPs with maf > 0.05\n",
      "dim for EUR file (before maf > 0.05)\n",
      "(503, 32504336)\n",
      "dim for EUR file (after maf > 0.05)\n",
      "(503, 7230618)\n",
      " \n",
      "filtering SNPs with mac > 1\n",
      "number of SNPs with mac > 1 (will be kept): 7230539\n",
      "dim for EUR file (after maf > 0.05, mac > 1)\n",
      "(503, 7230539)\n",
      "cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      "cf: dim original - (2504, 32504336)\n",
      " \n",
      "SnpData(people: 503, snps: 7230539,\n",
      "snp_info: \n",
      " Row │ chromosome  snpid         genetic_distance  position  allele1  allele2\n",
      "     │ String      String        Float64           Int64     String   String\n",
      "─────┼────────────────────────────────────────────────────────────────────────\n",
      "   1 │ 1           1:10177:A:AC               0.0     10177  AC       A\n",
      "   2 │ 1           1:10352:T:TA               0.0     10352  TA       T\n",
      "   3 │ 1           1:11008:C:G                0.0     11008  G        C\n",
      "   4 │ 1           1:11012:C:G                0.0     11012  G        C\n",
      "   5 │ 1           1:13110:G:A                0.0     13110  A        G\n",
      "   6 │ 1           1:13116:T:G                0.0     13116  G        T\n",
      "…,\n",
      "person_info: \n",
      " Row │ fid        iid        father     mother     sex        phenotype\n",
      "     │ Abstract…  Abstract…  Abstract…  Abstract…  Abstract…  Abstract…\n",
      "─────┼──────────────────────────────────────────────────────────────────\n",
      "   1 │ 0          HG00096    0          0          1          -9\n",
      "   2 │ 0          HG00097    0          0          2          -9\n",
      "   3 │ 0          HG00099    0          0          2          -9\n",
      "   4 │ 0          HG00100    0          0          2          -9\n",
      "   5 │ 0          HG00101    0          0          1          -9\n",
      "   6 │ 0          HG00102    0          0          2          -9\n",
      "…,\n",
      "srcbed: data_ancestry/kgp.EUR.maf005.mac1.bed\n",
      "srcbim: data_ancestry/kgp.EUR.maf005.mac1.bim\n",
      "srcfam: data_ancestry/kgp.EUR.maf005.mac1.fam\n",
      ")\n",
      " \n",
      " \n",
      "files to be removed: \n",
      "8\n",
      "Any[\"data_ancestry/kgp.EUR.bed\", \"data_ancestry/kgp.EUR.bim\", \"data_ancestry/kgp.EUR.fam\", \"data_ancestry/kgp.EUR.maf005.bed\", \"data_ancestry/kgp.EUR.maf005.bim\", \"data_ancestry/kgp.EUR.maf005.fam\", \"data_ancestry/kgp.EUR.maf005.mac1.fam\", \"data_ancestry/kgp.EUR.maf005.mac1.bim\"]\n",
      " \n",
      " \n",
      "### processing SAS samples ###\n",
      " \n",
      "### subsetting to SAS samples ###\n",
      "sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      " \n",
      "filtering SNPs with maf > 0.05\n",
      "dim for SAS file (before maf > 0.05)\n",
      "(489, 32504336)\n",
      "dim for SAS file (after maf > 0.05)\n",
      "(489, 7422918)\n",
      " \n",
      "filtering SNPs with mac > 1\n",
      "number of SNPs with mac > 1 (will be kept): 7422819\n",
      "dim for SAS file (after maf > 0.05, mac > 1)\n",
      "(489, 7422819)\n",
      "cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
      "cf: dim original - (2504, 32504336)\n",
      " \n",
      "SnpData(people: 489, snps: 7422819,\n",
      "snp_info: \n",
      " Row │ chromosome  snpid         genetic_distance  position  allele1  allele2\n",
      "     │ String      String        Float64           Int64     String   String\n",
      "─────┼────────────────────────────────────────────────────────────────────────\n",
      "   1 │ 1           1:10177:A:AC               0.0     10177  AC       A\n",
      "   2 │ 1           1:10352:T:TA               0.0     10352  TA       T\n",
      "   3 │ 1           1:11008:C:G                0.0     11008  G        C\n",
      "   4 │ 1           1:11012:C:G                0.0     11012  G        C\n",
      "   5 │ 1           1:13116:T:G                0.0     13116  G        T\n",
      "   6 │ 1           1:13118:A:G                0.0     13118  G        A\n",
      "…,\n",
      "person_info: \n",
      " Row │ fid        iid        father     mother     sex        phenotype\n",
      "     │ Abstract…  Abstract…  Abstract…  Abstract…  Abstract…  Abstract…\n",
      "─────┼──────────────────────────────────────────────────────────────────\n",
      "   1 │ 0          HG01583    0          0          1          -9\n",
      "   2 │ 0          HG01586    0          0          1          -9\n",
      "   3 │ 0          HG01589    0          0          1          -9\n",
      "   4 │ 0          HG01593    0          0          2          -9\n",
      "   5 │ 0          HG02490    0          0          1          -9\n",
      "   6 │ 0          HG02491    0          0          2          -9\n",
      "…,\n",
      "srcbed: data_ancestry/kgp.SAS.maf005.mac1.bed\n",
      "srcbim: data_ancestry/kgp.SAS.maf005.mac1.bim\n",
      "srcfam: data_ancestry/kgp.SAS.maf005.mac1.fam\n",
      ")\n",
      " \n",
      " \n",
      "files to be removed: \n",
      "8\n",
      "Any[\"data_ancestry/kgp.SAS.bed\", \"data_ancestry/kgp.SAS.bim\", \"data_ancestry/kgp.SAS.fam\", \"data_ancestry/kgp.SAS.maf005.bed\", \"data_ancestry/kgp.SAS.maf005.bim\", \"data_ancestry/kgp.SAS.maf005.fam\", \"data_ancestry/kgp.SAS.maf005.mac1.fam\", \"data_ancestry/kgp.SAS.maf005.mac1.bim\"]\n",
      " \n",
      " \n",
      "all ancestries were processed\n"
     ]
    }
   ],
   "source": [
    "ancestries = [\"AFR\", \"AMR\", \"EAS\", \"EUR\", \"SAS\"]\n",
    "for ancestry in ancestries\n",
    "    println(\"##### processing $(ancestry) samples #####\")\n",
    "    if isfile(\"data_ancestry/$(basefile).$(ancestry).maf005.mac1.bed\")\n",
    "        println(\"$(basefile).$(ancestry).maf005.mac1.bed exists; skip\")\n",
    "        println(\" \")\n",
    "        continue\n",
    "    end    \n",
    "\n",
    "    rm_files = []\n",
    "    kgp = kgp_original\n",
    "\n",
    "    #### Subset to the current ancestry samples ####\n",
    "    println(\"### subsetting to $(ancestry) samples ###\")\n",
    "    println(\"sample size ref: AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\")\n",
    "    ancestry_data = meta.sample[meta.super_pop .== \"$(ancestry)\"]\n",
    "\n",
    "    rowinds = findall(in(ancestry_data), kgp.person_info.iid)\n",
    "\n",
    "    file = \"$(basefile).$(ancestry)\"\n",
    "    SnpArrays.filter(kgp, rowinds, trues(size(kgp)[2]); des = \"data_ancestry/$(file)\")\n",
    "    println(\" \")\n",
    "\n",
    "\n",
    "\n",
    "    #### get colinds: maf > 0.05 ####\n",
    "    println(\"filtering SNPs with maf > 0.05\")\n",
    "    kgp = SnpData(\"data_ancestry/$(file)\") # the current ancestry samples: $(basefile).$(ancestry)\n",
    "    println(\"dim for $(ancestry) file (before maf > 0.05)\")\n",
    "    println(size(kgp))\n",
    "\n",
    "    file = \"$(basefile).$(ancestry).maf005\"\n",
    "    colinds = SnpArrays.filter(kgp.snparray; min_maf = 0.05)[2]\n",
    "    SnpArrays.filter(kgp, trues(size(kgp)[1]), colinds; des = \"data_ancestry/$(file)\")   \n",
    "\n",
    "    kgp = SnpData(\"data_ancestry/$(file)\") # current ancestry samples with maf > 0.05: $(basefile).$(ancestry).maf005\n",
    "    println(\"dim for $(ancestry) file (after maf > 0.05)\")\n",
    "    println(size(kgp))\n",
    "    println(\" \")\n",
    "\n",
    "\n",
    "\n",
    "    #### get colinds: MAC > 1 ####\n",
    "    println(\"filtering SNPs with mac > 1\")\n",
    "\n",
    "    # count matrix reference: https://openmendel.github.io/SnpArrays.jl/latest/#Counts\n",
    "    cnt = counts(kgp.snparray, dims = 1)\n",
    "\n",
    "    # removes all samples fall into one genotype, or all samples' genotype is missing\n",
    "    colinds = Int[]\n",
    "    for j in 1:size(kgp)[2]\n",
    "        size(kgp)[1] in cnt[:,j] ? nothing : push!(colinds, j)\n",
    "    end\n",
    "\n",
    "    snps_num_mac1 = length(colinds)\n",
    "    println(\"number of SNPs with mac > 1 (will be kept): $(snps_num_mac1)\")\n",
    "\n",
    "    # for each column, check if any of the rows have all samples\n",
    "    # ref. AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\n",
    "    # if none of the rows == above ancestry sample size in the column, the column will be kept\n",
    "    file = \"$(basefile).$(ancestry).maf005.mac1\"\n",
    "    SnpArrays.filter(kgp, trues(size(kgp)[1]), colinds; des = \"data_ancestry/$(file)\")\n",
    "\n",
    "\n",
    "    kgp_filtered = SnpData(\"data_ancestry/$(basefile).$(ancestry).maf005.mac1\")\n",
    "    size_kgp_filtered = size(kgp_filtered)\n",
    "    println(\"dim for $(ancestry) file (after maf > 0.05, mac > 1)\")\n",
    "    println(size_kgp_filtered)\n",
    "    println(\"cf: sample size ref - AFR 661, AMR 347, EAS 504, EUR 503, SAS 489\")\n",
    "    println(\"cf: dim original - $(size_kgp_original)\")\n",
    "    println(\" \")\n",
    "\n",
    "    println(kgp_filtered)\n",
    "    println(\" \")\n",
    "    println(\" \")\n",
    "\n",
    "\n",
    "    #### clean up the directory; removing unncessesary ancestry files ####       \n",
    "    rm_file_list = filter(x -> !occursin(r\"mac\", x), glob(\"$(basefile).$(ancestry)*\", \"data_ancestry\"))\n",
    "    for file in rm_file_list\n",
    "        rm_files = push!(rm_files, file)\n",
    "    end\n",
    "    rm_files = push!(rm_files, \"data_ancestry/$(basefile).$(ancestry).maf005.mac1.fam\")\n",
    "    rm_files = push!(rm_files, \"data_ancestry/$(basefile).$(ancestry).maf005.mac1.bim\")\n",
    "\n",
    "    println(\"files to be removed: \")        \n",
    "    println(length(rm_files))\n",
    "    println(rm_files)        \n",
    "\n",
    "    for file in rm_files\n",
    "        rm(file)\n",
    "    end\n",
    "    println(\" \")\n",
    "    println(\" \")\n",
    "end\n",
    "println(\"all ancestries were processed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.8.1",
   "language": "julia",
   "name": "julia-1.8"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
